### DISCUSSION TITLE
dnsmasq - Vorschläge

### DISCUSSION THREAD

**Discussion by flosch-dev** [Freetz-NG/freetz-ng#258] [Allgemein]
Damit das nicht im PR alles diskutiert wird, wollte ich mal hier anfangen.
Ich würde gerne 3 weitere Veränderungen am dnsmasq package vorschlagen:
### DNS Server:
* option to use "all-servers" 
By default, when dnsmasq has more than one upstream server available, it will send queries to just one server. Setting this flag forces dnsmasq to send all queries to all available servers. The reply from the server which answers first will be returned to the original requester.

### DHCP Server:
* option to disable dhcp event logging "quiet-dhcp"
Suppress logging of the routine operation of these protocols. Errors and problems will still be logged. --quiet-dhcp and quiet-dhcp6 are over-ridden by --log-dhcp.

#### Generel:
* display current DHCP leases
Ich habe bei mir mal am Ende von dnsmasq.cgi einfach folgendes hinzugefügt:
```
sec_begin "$(lang de:"DHCP leases" en:"DHCP leases")"
cat << EOF
<h2>$(lang de:"Aktuelle DHCP leases:" en:"Current dhcp leases:"):</h2>
EOF
echo -n '<pre id="log" class="log full"><FONT SIZE=-1>'
awk 'NF>1{ $1=strftime("%c", $1);}1' /tmp/dnsmasq.leases | html
echo '</FONT></pre>'
sec_end
```

Ich gebe zu, dass die DNS und DHCP parameter sicherlich auch einfach in der extra eingetragen werden können, aber ich denke das sind sinnvolle Parameter die sicher einige nutzen würden.

Bezüglich der Leases, ich bin nicht sicher ob die Position (am Ende von dnsmasq.cgi) die beste ist, aber ich habe leider nicht rausgefunden, wie ich auf die schnelle ein weiteres cgi hinzufügen kann, welches dann bei dnsmasq auftaucht. Bevor ich aber weiter daran arbeite, wollte ich mal Feedback einholen.

Beste Grüße

**Comment by fda77** [comment]
Ich weiss nicht, ich bin nicht der typische Durrchschnittsbenutzer. "all-servers" hab ich eh schon in den extras.  "quiet-dhcp" würde ich nicht einschalten. "display current DHCP leases" finde ich praktisch, das könnte man wie http://fritz:81/cgi-bin/file/inetd/conf machen. Diese Art cgi braucht nur die Datei die angezeigt werden soll. Das sollte dann aber auch nur sichtbar sein wenn der DHCP überhaupt an ist. Da wäre es wohl am besten in die .save den Aufruf von "/mod/etc/init.d/rc.dnsmasq regcgi" zu schreiben. Dieses regcgi in der rc würde dann eine interne Funktion aufrufen die auch vom load aufgerufen wird. Dann braucht man die modunreg+modreg nur in diese Funktion 1x zu schreiben

**Comment by flosch-dev** [comment]
OK, ich denke die parameter sind vielleicht doch besser in der extra aufgehoben, das erleichter auch das maintainen falls sich daran was ändert.
Bezüglich der Leases Anzeige mach ich mal einen Vorschlag.
- seperate Seite mit formatierter Ausgabe der current leases (expire Zeitstempel lesbar)
- conditional wenn DHCP Server aktiviert

**Comment by fda77** [comment]
Ne komplette cgi brauchst du nicht zu machen

Entwedet eine DATEI  read-only anzeigen: 
https://github.com/Freetz-NG/freetz-ng/blob/master/make/inetd/files/root/etc/default.inetd/inetd_conf.def

oder einen BEFEHL der die Ausgabe generiert:
https://github.com/Freetz-NG/freetz-ng/blob/master/make/lcd4linux/files/root/etc/default.lcd4linux/lcd4linux_list.def

**Comment by flosch-dev** [comment]
Ich habe ein .def file mit folgendem CONFIG_CMD angelegt:
```
CONFIG_CMD="awk 'NF>1{ \$1=strftime(\"%c\", \$1);}1' /tmp/dnsmasq.leases"
```
Die maskierung der quotes müsste passen:

```
root@fritz.box:/var/media/ftp# CONFIG_CMD="awk 'NF>1{ \$1=strftime(\"%c\", \$1);}1' /tmp/dnsmasq.leases"
root@fritz.box:/var/media/ftp# echo $CONFIG_CMD
awk 'NF>1{ $1=strftime("%c", $1);}1' /tmp/dnsmasq.leases
```

Leider bleibt die Seite bzw. das Textfeld aber immer leer. Jetzt habe ich mal ein file "get_leases" mit folgendem Inhalt angelegt:
```
#!/bin/sh
if [ -f "/tmp/dnsmasq.leases" ]; then
  awk 'NF>1{ $1=strftime("%c", $1);}1' /tmp/dnsmasq.leases
else
  echo "no leases file found"
  exit 1
fi
```
und wenn ich nur dieses in CONFIG_CMD eintrage `"CONFIG_CMD="/var/media/ftp/get_leases"` scheint es zu funktionieren.

Wäre das in Ordnung, das file get_leases mit anzulegen?

**Comment by fda77** [comment]
Hab hab es auch nicht hinbekommen CONFIG_CMD richtig zu füllen. Das muss irgendwas mit escaping und shell-expansion zu tun haben
Statt einer eigenen "get_leases" Datei könntest du der rc.dnsmasq auch ein eine parameter "get_leases" hinzufügen. Der Test "-s" prüft übrigens ob es eine Datei gibt und diese größer 0 Bytes ist, falls das hier sinnvoll ist

**Comment by flosch-dev** [comment]
Danke, eine funktion im rc ist sicher viel besser, gefällt mir. Ich probiere das mal aus.

**Comment by flosch-dev** [comment]
Eine weitere Idee die ich gerne Besprechen möchte, dass in der WebGUI für jedes package auch die versions nummer angezeigt wird.
Ich habe mir mal die Anzeige vom hosts-file bei addhole.cgi abgeguckt. (glaube das stammt von dir?)
Als Beispiel für dnsmasq könnte das so aussehen:
```
sec_begin "Package Version"
echo -n '<pre style="width:650px;overflow:auto"><FONT SIZE=-1>'
dnsmasq -v | html
echo '</FONT></pre>'
sec_end
```

![image](https://user-images.githubusercontent.com/47741987/115152867-4983f500-a073-11eb-8038-5ffb72224974.png)

Also ich fänd das halt für die meißten packages schon sinnvoll zu sehen welche daemon version vorhanden ist. Man kann das sicher auch über die Hilfe herausfinden, aber im Fall von dnsmasq gibt es ja dort 2 version und man wird auch immer auf die aktuelle freetz-ng man page gelinkt, auch wenn man eine alte Version bentutz.

**Reply by fda77** [reply]
So funktioniert das aber nur für diese 1 Package. Bei anderen wird die Version evtl gar nicht angezeigt oder man braucht andere Parameter. Man könnte es für alle Packages machen, zb dass die Version im Kopf angezeigt wird. Es gibt bereits eine /etc/static.pkg in der aber keine Versionen sind. Man könnte noch build/modified/firmware/var/.packages mit ins Image nehmen und die Versionen dort auslesen

**Comment by flosch-dev** [comment]
wie wird denn build/modified/firmware/var/.packages erstellt? Könnte man ein anderes Trennzeichen als "-" benutzen? 
Eine Lösung die für alle Pakete funktioniert ist sicherlich besser.
Ich bin nicht ganz vertraut mit dem freetz webUI, aber müsste man dass dann im "skin.sh" jeweils für alle skins anpassen?

**Comment by fda77** [comment]
Die wird in der fwmod erstellt, ein anderer Trennen ist viel Aufwand. Kannst ja mal schauen. Wenn man ncith jedes Skin ändern will könnte man auch die List auf  "Box-info" erweitern

**Comment by fda77** [comment]
https://github.com/Freetz-NG/freetz-ng/commit/0a88bfe134e8eddabfbb632b26f22f97ed031d44

zb
```
PKG=haserl
VER="$(sed -n "s/^$PKG //p" /etc/versions.pkg)"
echo $VER
```

VER kann  leer sein

**Comment by flosch-dev** [comment]
Danke für den commit, damit lässt sich schon was anfange.
Was hälst du von folgendem commit für Mod-Info: dac9ca8360e71dc05ec1a422d74fa92174eee5b8 ?

ca_bundle wird irgendwie ca bundle in /etc/versions.pkg benannt, aber in mod-Info halt mit "_".
Müsste man das im package entsprechend ändern?

Ich habe in dem branch auch noch die subtypes etwas eingerückt (3x space), damit lassen sie die main packages besser überblicken. Das ist ein anderer commit.

**Comment by fda77** [comment]
Mod-info nutzt die Variablennamen, und die dürfen keinen Bindestrich enthalten. In der Datei sind aber die richtigen Packagenamen.
Schau mal ob das funktioniert. Einfach für sed einen Platzhalter. Evtl muss man die 3 "/" von sed durch was anderes Ersetzen, zb ","
```
ver="$(sed -n "s/^$entry //p" /etc/versions.pkg)"
ver="$(sed -n "s/^${entry//-/.} //p" /etc/versions.pkg)"
```
Es kann sein dass für andere Packages noch mehr Ersetzungen gebraucht werden. Dafür gibt es kein einheitliches System in Freetz.

Manche Versionen sind lang, zb juis_check 32 Zeichen vom git-hash. Man könnte die Zeichenanzahl begrenzen:
```
echo "$name $ver<br>"
echo "$name ${ver:0:10}<br>"
```

(alles ungetestet)

**Comment by flosch-dev** [comment]
wieder mal Danke! 
Der Tipp war sehr gut und funktioniert. - 6814e18c3c6fdb68adbb021cbd4354e5d668a410
Zumindest bei mir für ca_bundle und juis_check. Weißt gerade nicht ob es noch andere cases gibt, aber die version würde dann halt erstmal nur nicht angezeitg.
Auch das Kürzen auf 10 Zeichen finde ich eine gute Idee. Das fiel mir auf Grund des "_" Problems noch gar nicht auf.

![image](https://user-images.githubusercontent.com/47741987/116797291-870d6700-aae4-11eb-834f-6a250821153e.png)

**Comment by flosch-dev** [comment]
Ich hatte leider noch $pkg drin, das war aber zum testen. 
Also der change würde aktuell so aussehen:
https://github.com/Freetz-NG/freetz-ng/compare/master...flosch-dev:mod-info_version

Für alle Packages in meinem build werden die versionsnummer damit angezeigt.

**Reply by fda77** [reply]
Ich bin mit der versions-Datei noch nicht ganz zufrieden, zb "syslog cgi-1.2.3". Das ist aber unabhängig vom mod-info

**Reply by flosch-dev** [reply]
Hm, das könnte man im fwmod oder halt info rausfiltern. Denkst du besser in fwmod rausfiltern, so dass versions.pkg kein cgi mehr enthält?

**Reply by fda77** [reply]
Ich hab die Datei angepasst, jetzt sollte alles korrekt drin stehen, jetzt ohne rückwärts-parsen: https://github.com/Freetz-NG/freetz-ng/commit/196abfa559c3ba7a6a6aa2386f5547b7a19f5781
Du musst deinen Änderung aber nochmal anpassen:
```
        else
-               echo "$name<br>"
+               [ "$type" == "cgi" ] && name="$name-cgi"
+               [ "$type" == "cgi" -o "$type" == "pkg" ] && ver="$(cut /etc/packages.lst -d ' ' -f 3,4 | sed -n "s/^${name//_/.} / /p")" || ver=''
+               echo "$name${ver:0:13}<br>"
                open_entry=$name
        fi
```
Die Seite lädt nun viel schneller, da nur für die Packages eine Version gesucht wird, ich hab viele Module im Image. Die Länge hab ich auf 13 gesetzt, damit die wireguard Versionen nicht abgeschnitten werden

**Comment by flosch-dev** [comment]
wow, jetzt hattest du doch ganz schön viel Arbeit. Hättest auch ruhig den change für Mod-Info mit commiten können, der Inhalt ist ja auch zum Großteil deine Arbeit bzw dein Verdienst.
Aber freue mich, dass die Version dann auch in der Package page mit auftaucht. Ich baue das heute Abend gleich noch und probiere es einmal bei mir aus.
Danke!

**Reply by fda77** [reply]
Wenn man Sachen andere committed hat, gab es frühe ab und an mal Ärger, daher mach ich das nicht so gern. Mache ich dann gleich noch, ich hab das schon so laufen
Der erste einfach Versuch mit der versions.pkg hätte zu viele Workaround nach sich gezogen. So gibts ne Datei die auch von anderen Komponenten aufgerufen werden könnte, mit der Version wie sie in der .mk steht

**Reply by fda77** [reply]
Die Version hab ich jetzt direkt in der packages.lst verkürzt. Nun 12 Stellen damit auch Wireguard korrekt angezeigt wird

**Reply by fda77** [reply]
-> 807d33ce3 (und 7fd2f5142 ...)

**Reply by flosch-dev** [reply]
Da du die meißte Arbeit rein gesteckt hast, sehe ich kein Problem wenn du den commit erstellst.
Das mit der header Zeile im skin ist auch super! Danke für die Mühe, ich finde das hast sich gelohnt. Bisher musste ich immer manuell die daemons prüfen um zu sehen welche version läuft.

Ich habe nochmal etwas an dem output für die DHCP leases gefeilt, geht das so in Ordnung? https://github.com/Freetz-NG/freetz-ng/pull/287

**Reply by fda77** [reply]
Mach einfach, ich will mich gar nicht um alles kümmern ;-) Da ich mehrere DHCP verwende, kann ich das eh nicht sinnvoll nutzen, da auf keinem Server alle Hosts registriert sind

**Comment by flosch-dev** [comment]
Habe heute Abend noch ein weiteres kleines Projekt gestartet, aber bräuchte da mal etwas Rat für die weitere Umsetzung.
Ich habe mir auf die Box ein file mit allen MAC Address Vendoren https://gitlab.com/wireshark/wireshark/-/raw/master/manuf geladen und möchte dieses nutzen, um den Vendor zusätzlich in den DHCP leases anzuzeigen.
Das File ist ca 1.8 MB groß und das parsen entsprechend langsam.

- wie kriegt man dieses file  ggf. kompriemiert in das image? tar.gz ist ca 551 kbytes
- Da das parsen sehr langsam ist, wäre es möglich beim Aufruf von dnsmasq_leases.def noch eine GET oder POST var dranzuhängen um das parsen so manuel zu starten? Also das per default das leases file wie jetzt angezeigt wird und es einen button mit "show mac vendor" gibt.
- Sollte ich am besten ein package mac-vendor mit dem vendor file + shell script erstellen oder mit ins dnsmasq? Man könnte ja eine option "mac vendor lookup" in Menuconfig unter dnsmasq einfügen?

Das parsing läuft über ein file:
```
#!/bin/sh
#
VENDORFILE="/var/media/ftp/manuf"

if [ $(echo $1 | grep -Eo '([[:xdigit:]]{1,2}:){5}[[:xdigit:]]{1,2}') ]; then
  VENDOR=${1:0:8}
  grep -i $VENDOR $VENDORFILE | awk '{printf $3" "$4}'
else
  echo "invalid MAC: $1"
  exit 1
fi
```

und dann der Aufruf:

` grep -E '[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}' /tmp/dnsmasq.leases | sort -n -t . -k 2,2 -k 3,3 -k 4,4 | awk 'NF>1{ $1=strftime("%Y-%m-%d %H:%M:%S", $1);}1' | awk '{printf $0" ("; system("/var/media/ftp/lookup_mac.sh "$3);print")"}'`

**Reply by flosch-dev** [reply]
Zum Beschleunigen könnte man im 1. Schritt ein temp mac vendor file erstellen wo alle mac aus dem lease file drin sind. also einmal grep mit allen macs nach tmp.
Danach parsed man nur noch das tmp file, das geht schneller

**Comment by fda77** [comment]
Als eigenes Package ist auf jeden Fall einfacher. Gibt es diese Liste nicht schon von einem anderen Package, zb bind/tcpdump/nmap? Die sind aber nicht so aktuell!
Um auf der Fritzbox so viele Einträge schneller zu durchsuchen könnte man mehrere Dateien machen, benannt nach dem 1 oder 2 ersten Bytes/Zeichen. Das ist schonmal um den Faktor 16 bzw 256 schneller...
Oder du lässt das erstmalige Suchen langsam wie es ist, und legst ein Cache unter /tmp/flash/PKG/FILE an. Bei den meisten Benutzer sind es nach einem Reboot noch die gleichen Hersteller
Um die Suche schneller zu machen sollte man aber "grep -i" vermeiden, das sind unnötig viele Vergleiche, 6 Byte/Zeichen sind 64 Kombinationen, und nur ab Zeilenanfang (^). In der Datei ist eh alles gross geschrieben! Besser nur die Variable gross machen: 'grep "^${VENDOR^^}" $VENDORFILE'
EDIT: Die Busybox auf der Fritzbox kann  scheinbar kein '${x^^}'
